src/main/scala/com/bazaarvoice/sswf/model/StepsHistory.scala

```diff
@@ -0,0 +1,171 @@
+import com.amazonaws.services.simpleworkflow.model._
+import com.bazaarvoice.sswf.StepEventState._
+import com.bazaarvoice.sswf.{InputParser, SSWFStep, StepEventState}
+import org.joda.time.{DateTime, Duration}
+
+import scala.collection.JavaConversions._
+import scala.collection.mutable
+import scala.reflect.{ClassTag, classTag}
+
+object WorkflowEventToken {
+  override def toString = "WORKFLOW"
+}
+
+case class StepEvent[StepEnum <: (Enum[StepEnum] with SSWFStep) : ClassTag](id: Long,
+                                                                            event: Either[StepEnum, WorkflowEventToken.type],
+                                                                            result: String,
+                                                                            state: StepEventState,
+                                                                            start: DateTime,
+                                                                            end: Option[DateTime],
+                                                                            cumulativeActivityTime: Duration)
+
+case class StepsHistory[SSWFInput, StepEnum <: (Enum[StepEnum] with SSWFStep) : ClassTag](input: SSWFInput, events: java.util.List[StepEvent[StepEnum]], firedTimers: java.util.Set[StepEnum])
+
+object HistoryFactory {
+  def from[SSWFInput, StepEnum <: (Enum[StepEnum] with SSWFStep) : ClassTag](swfHistory: List[HistoryEvent],
+                                                                             inputParser: InputParser[SSWFInput]): StepsHistory[SSWFInput, StepEnum] = {
+
+    def enumFromName(name: String): StepEnum = Enum.valueOf(classTag[StepEnum].runtimeClass.asInstanceOf[Class[StepEnum]], name)
+
+    def filterStarts(s: List[StepEvent[StepEnum]]): List[StepEvent[StepEnum]] = s match {
+      case Nil | _ :: Nil                                   =>
+        s
+      case first :: second :: rest if first.id == second.id =>
+        filterStarts(rest)
+      case first :: rest                                    =>
+        first :: filterStarts(rest)
+    }
+
+    val steps = mutable.Map[Long, HistoryEvent]()
+    var startTime: DateTime = null
+    val startTimes = mutable.Map[String, DateTime]()
+    var input: String = null
+
+    val startedTimers = mutable.Map[String, HistoryEvent]()
+    val firedTimers = mutable.Set[String]()
+    val cancelledTimers = mutable.Set[String]()
+    var completedStepCounts = Map[StepEnum, Int]().withDefaultValue(0)
+
+    val eventsRaw =
+      for {h <- swfHistory} yield {
+        EventType.fromValue(h.getEventType) match {
+          case WorkflowExecutionStarted   =>
+            startTime = new DateTime(h.getEventTimestamp)
+            input = h.getWorkflowExecutionStartedEventAttributes.getInput
+            Some(StepEvent[StepEnum](h.getEventId, Right(WorkflowEventToken), "STARTED", STARTED, startTime, None, new Duration(startTime, startTime)))
```

Comment:
Is there any difference between `new Duration(startTime, startTime)` and `Duration.ZERO`?